### DISCUSSION TITLE
How to get Module-Build-Tiny-0.048 to pass all of its tests.

### DISCUSSION THREAD

**Discussion by sisyphus** [StrawberryPerl/Perl-Dist-Strawberry#210] [General]
I note that SP-5.40.0.1-PDL builds Module-Build-Tiny-0.048 just fine, with all tests passing.
(There's a whole lot of junk at the end about temp files not being cleaned up, but that's not really an issue.)

Whenever I try to build the module, the final test ('Script is installed') always fails (as per https://github.com/Perl-Toolchain-Gang/module-build-tiny/issues/35).

I've spent a couple of hours trying to work out what SP has done that enables all tests to pass ... but haven't yet found the key.

Does anyone have any tips ?

**Comment by shawnlaffan** [comment]
It's not built directly so must be done as a prereq for another module.  

Possibly that is run with the ```ignore_testfailure``` flag set (which I assume propagates to dependencies but I have not checked).  
https://github.com/StrawberryPerl/Perl-Dist-Strawberry/blob/master/share/64bit-5.40.0.1.pp

Edit - 

Although that's not what you're reporting...

What about the various config settings that control script installation location and the like?

**Reply by sisyphus** [reply]
> What about the various config settings that control script installation location and the like?

Yeah ... maybe. (I compared the environment settings, and found no unexpected difference except for a <code>HOME</code> setting, to which I attended.)

5.40.0.1-PDL does actually successfully build M-B-T-0.048 (and pass all tests)  when I do a manual build of that module.
But all of my other perl-5.40.0 builds will fail the final test, as per https://github.com/Perl-Toolchain-Gang/module-build-tiny/issues/35.

I've built a quadmath perl-5.40.0 from source, using the toolchain provided by SP 5.40.0.1-PDL, and set it up to use the 3rd party libraries provided by that SP build.
I'm curious to see how those libraries (and all of the perl modules also) go with a quadmath build of perl ... and I certainly don't want to go to the trouble of building all of those libraries myself ;-)

Having built that perl, I've started working through the list provided in DISTRIBUTIONS.txt, and leaving the "Alien" stuff till last,
I'm being careful to install the same version as specified in DISTRIBUTIONS.txt, but I've hit this anomaly with M-B-T-0.048 when  it came to installing App-pmuninstall.
I don't think this is "quadmath" issue.

It's not a big deal, and there's no hurry for me to complete the task, so I'll just keep pushing on .... and I'll try to keep my "pestering" to a minimum.

I think it's worthwhile understanding how it has come about that SP-5.40.0-PDL avoids this issue with M-B-T-0.048, while so many other builds don't - so I'll leave this open.

**Comment by shawnlaffan** [comment]
Is it perhaps related to the architectured setup described in https://github.com/Perl-Toolchain-Gang/ExtUtils-MakeMaker/issues/459#issuecomment-2296545743?

**Reply by sisyphus** [reply]
No - not related to that architecture.
This latest build of mine is laid out the same as SP-5.40.0.1-PDL. (But no "vendor" folder, of course.)
To make it a bit simpler, I've now built the same configuration of perl-5.40.0 except that nvtype is double instead of __float128.
I still see the same failure with t/simple.t.

One thing that puzzles me is that SP-5.40.0.1-PDL claims that Win32::Pipe is at version 0.025:
```
>perl -MWin32::Pipe -le "print $Win32::Pipe::VERSION"
0.025
```
And DISTRIBUTIONS.txt agrees with that.
But the only source I can find for that is https://cpan.metacpan.org/authors/id/J/JD/JDB/Win32-Pipe-0.025.tar.gz
and when I unpack that source, I find that SP-5.40.0.1-PDL cannot build it.
```
Cpipe.cpp:209:36: error: invalid conversion from 'char' to 'LPSTR' {aka 'char*'} [-fpermissive]
  209 |         szError[ERROR_TEXT_SIZE] = '\0';
      |                                    ^~~~
      |                                    |
      |                                    char
gmake: *** [makefile:340: Cpipe.o] Error 1
```
Nor can I get it to build.
So I'm a bit puzzled as to how Win32-Pipe-0.025 has been built and installed.
OTOH, I've installed version 0.026 on my builds without any problem.
(Win32::Pipe is part of the Module::Build::Tiny dependency chain.)

However, I don't think any of that is related to the issue I'm seeing.
If I update SP-5.40.0.1-PDL to Win32-Pipe-0.026, t/simple.t still passes all tests - and that suggests that the Win32-Pipe version is not important here.

BTW, the exact failure I'm seeing is that the final test looks for "install/bin/simple" but the "install" folder does not even contain a folder named "bin".

The second last test looks for and successfully finds "install\lib\perl5\MSWin32-x64-multi-thread\Foo\Bar.pm" on my "double" build (or "install\lib\perl5\MSWin32-x64-multi-thread-quadmath\Foo\Bar.pm" on my "quadmath" build).

Oh well .... something for tomorrow ;-)

**Reply by shawnlaffan** [reply]
Win32::Pipe uses a patched version.  I need to update that for the next build.  
https://github.com/StrawberryPerl/Perl-Dist-Strawberry/blob/8ec82760ac883c7a6a00425fc2ec53967526b528/share/64bit-5.40.0.1.pp#L205

**Reply by shawnlaffan** [reply]
Looking at the test, MBT generates a dummy distribution using `t/lib/DistGen.pm`, changes directory to it, and then populates it with files.  

So the `script` file should live under something like `./EjtxKHSwqX/Foo-Bar/script`, where the `Ejtx...` is a temp dir name.

**Reply by sisyphus** [reply]
I've checked that the version numbers of the modules loaded by DistGen.pm match. There's no difference there.

I modified t/simple/t to include some diagnostic info amongst the final 2 tests:
```
....
  warn "\nCWD: ", Cwd::getcwd(), "\n"; # DistGen.pm has loaded the Cwd module
  warn "FILENAME: $filename\n";
  ok( -f $filename, 'Module is installed');
  my $dir = `dir "install"`;
  warn "DIR: $dir\n";
  ok( -f 'install/bin/simple', 'Script is installed'); 
```
For my quadmath build, when I run <code>perl t/simple.t</code> that produces:
```
.....
ok 30 - Build install output looks correctly

CWD: D:/s/Module-Build-Tiny-0.048/9cXdqohYPU/Foo-Bar
FILENAME: install\lib\perl5\MSWin32-x64-multi-thread-quadmath\Foo\Bar.pm
ok 31 - Module is installed
DIR:  Volume in drive D is Data
 Volume Serial Number is 8252-9DA9

 Directory of D:\s\Module-Build-Tiny-0.048\9cXdqohYPU\Foo-Bar\install

03/09/2024  11:02 AM    <DIR>          .
03/09/2024  11:02 AM    <DIR>          ..
03/09/2024  11:02 AM    <DIR>          lib
               0 File(s)              0 bytes
               3 Dir(s)  970,219,261,952 bytes free

not ok 32 - Script is installed
#   Failed test 'Script is installed'
#   at t/simple.t line 172.
1..32
```
Running the same script (from the same location) using SP-5.40.0.1-PDL, produces:
```
ok 30 - Build install output looks correctly

CWD: D:/s/Module-Build-Tiny-0.048/3EDciZR7jW/Foo-Bar
FILENAME: install\lib\perl5\MSWin32-x64-multi-thread\Foo\Bar.pm
ok 31 - Module is installed
DIR:  Volume in drive D is Data
 Volume Serial Number is 8252-9DA9

 Directory of D:\s\Module-Build-Tiny-0.048\3EDciZR7jW\Foo-Bar\install

03/09/2024  11:25 AM    <DIR>          .
03/09/2024  11:25 AM    <DIR>          ..
03/09/2024  11:25 AM    <DIR>          bin
03/09/2024  11:25 AM    <DIR>          lib
               0 File(s)              0 bytes
               4 Dir(s)  970,219,216,896 bytes free

ok 32 - Script is installed
1..32
```
That all looks fine and consistent to me, except for the absence of the "bin" directory in the output of the quadmath build's <code>dir</code> command.

I just need to find out what's supposed to generate that "bin" directory (and its contents), and why my perl installation is failing to do so.

**Reply by shawnlaffan** [reply]
What happens if you run `dir /s simple | findstr Directory` under the generated temp dir?  

Set CLEANUP to 0 first if not already done: https://github.com/Perl-Toolchain-Gang/module-build-tiny/blob/a59c6833a457305353090cd3bb8368609ac3dfaf/t/lib/DistGen.pm#L54

I get three locations containing the file `simple`.

```
...\Foo-Bar\blib\script
...\Foo-Bar\install\bin
...\Foo-Bar\script
```

If I understand correctly, the file is created in `script`, copied to `blib\script` and then installed to `install\bin`.

(Edited final location to include bin)

**Reply by sisyphus** [reply]
Yes, for both perls, I see a file named 'simple' in <code>Foo-Bar\script</code> and 2 files ('simple' and 'simple.bat') in <code>Foo-Bar\blib\script</code> - but it's only the Strawberry Perl build that provides the <code>Foo-Bar\install\bin</code> folder (containing 'simple' and 'simple.bat').

**Reply by shawnlaffan** [reply]
MBT uses ExtUtils::Installpaths, which in turn uses ExtUtils::Config.  

This set looks relevant: https://metacpan.org/pod/ExtUtils::InstallPaths#installdirs

Maybe add some debugging to MBT to see where it thinks things should be going?  Maybe here: https://metacpan.org/release/LEONT/Module-Build-Tiny-0.049/source/lib/Module/Build/Tiny.pm#L146

**Reply by sisyphus** [reply]
Yes - for both my own quadmath build and Strawberry's build, I had changed that piece of code to:
```
	install => sub {
		my %opt = @_;
                use Data::Dump qw(dump);
                dump(%opt);
		die "Must run `./Build build` first\n" if not -d 'blib';
		install($opt{install_paths}->install_map, @opt{qw/verbose dry_run uninst/});
		return 0;
	},
```
but hadn't got around to fully evaluating the diff of the 2 dumps.
Anyway, this morning I noticed:
```
-                                           script  => "C:\\sp\\_64\\sp-5.40.0.1-PDL\\perl\\bin",
+                                           script  => "",
```
which led me to realize that $Config{installsitescript} is unset in all of my builds of perl - and (as I've now established) that's why the test fails.

Right at the start you suggested "What about the various config settings that control script installation location and the like?".
That seemed a reasonable suggestion at the time, but I was thinking more about "installsitebin" and failed to notice the absence of a spec for "installsitescript".
It's so obvious now, and I apologise (again) for my lack of lucidity and thoroughness.

I feel this is a bug in perl - surely "installsitescript" should default to "installsitebin" unless some other (non-empty) value is specified ?
Both "installbin" and "installscript" point to the same location (ie "perl/bin"), and I've never had to specify either of those settings in the GNUmakefile args.

<strike>I'm thinking this should be addressed in perl's <code>win32/GNUmakefile</code> and <code>win32/Makefile</code> (as MSVC builds also specify the empty string for "installsitescript"). </strike>

UPDATE: A quick look indicates that it would be fixable in <code>win32/config.gc</code> and <code>win32/config.vc</code>, not in the makefiles as I initially thought.

Or, should it be addressed in EU::MM ?

This should fix https://github.com/Perl-Toolchain-Gang/module-build-tiny/issues/35

I'll leave this open until I've prepared a PR ... or until someone can satisfy me that it's not a perl bug.

**Reply by shawnlaffan** [reply]
All good.  It's useful to get to the bottom of these things, and now we have some searchable content for others to find.  

WRT defaulting `installscript` to `installbin`, possibly all three locations would be good.  Other perl builds will also not set this, and EU::MM will be used with them.

**Reply by shawnlaffan** [reply]
Actually, the change should perhaps be in ExtUtils::Config instead of EU::MM as that is what they all(?) seem to use.

**Reply by sisyphus** [reply]
Seems to me that Windows is probably the only system where builds of perl from "straight-out-of-the-box" perl source will invariably leave "installsitescript" as the empty string.
On Ubuntu, MSYS2, and Cygwin the system perls, and my own personal builds, all set "installsitescript" to a sensible value.

On the basis of what I see them doing, I think I'll prepare a PR (to <code>win32/config.gc</code> and <code>win32/config.vc)</code> in the next day or two, such that Windows perl builds will set "installsitescript" to the same string as "installscript".

However, given that we now know that there are perl builds around that have "installsitescript" set to the empty string, it's probably a good idea that EU::MM and/or EU::Config are also patched to handle that scenario.

Do you know why it is that Strawberry Perl wants/needs to customize the setting of these values in the way it does ?
I mean, why can't these 4 "Install..." options all just specify "Strawberry_path\perl\bin" ?

**Reply by shawnlaffan** [reply]
> Do you know why it is that Strawberry Perl wants/needs to customize the setting of these values in the way it does ?
> I mean, why can't these 4 "Install..." options all just specify "Strawberry_path\perl\bin" ?

It all pre-dates my work on the project.  

https://github.com/search?q=repo%3AStrawberryPerl%2FPerl-Dist-Strawberry%20installsitescript&type=code

My change on this front was to collate the changes into one place in 78adc57dadf47c4850cce44b0d501714dac23f8a to avoid having to patch Config.gc for every build.

**Reply by shawnlaffan** [reply]
Hmm.  That last response does not answer the question as posed.  

It still pre-dates me, but it makes sense to keep the core perl stuff separate from the vendor-built stuff (done as part of the SP build process) from those built by users themselves.  There is less risk that way of overwriting critical components when a module update goes wrong.

**Reply by sisyphus** [reply]
Oh, yeah - seems reasonable.
I'll be suggesting this change to win32/config.gc:
```
--- perl-5.41.3/win32/config.gc 2024-07-25 00:58:48.000000000 +1000
+++ perl-5.41.31/win32/config.gc        2024-09-05 13:15:13.128424900 +1000
@@ -849,7 +849,7 @@
 installsitelib='~INST_TOP~\site~INST_VER~\lib'
 installsiteman1dir=''
 installsiteman3dir=''
-installsitescript=''
+installsitescript='~INST_TOP~~INST_VER~\bin'
 installstyle='lib'
 installusrbinperl='undef'
 installvendorarch=''
```
I take it that change would have no impact at all on Strawberry perl builds.

**Reply by shawnlaffan** [reply]
Looks good to me

**Comment by sisyphus** [comment]
I later realized that installscript and installsitescript should honour <code>\~INST_ARCH\~</code> as is  done by installbin and installsitebin.
Hence: https://github.com/Perl/perl5/pull/22581

**Comment by sisyphus** [comment]
https://github.com/Perl/perl5/pull/22581 has been merged.
